pkg/agent/route/route_linux.go

```diff
@@ -1163,21 +1163,42 @@ func (c *Client) AddClusterIPRoute(svcIP net.IP) error {
 			} else {
 				c.clusterIPv4CIDR = newClusterIPCIDR
 			}
-			klog.V(4).InfoS("Created a route with new ClusterIP CIDR to route the ClusterIP to Antrea gateway", "ClusterIP CIDR", newClusterIPCIDR, "clusterIP", svcIP)
+			klog.V(4).InfoS("Created a route with new ClusterIP CIDR to route the ClusterIP to Antrea gateway", "ClusterIP CIDR", newClusterIPCIDR, "ClusterIP", svcIP)
 		} else {
-			klog.V(4).InfoS("Route with current ClusterIP CIDR can route the ClusterIP to Antrea gateway", "ClusterIP CIDR", curClusterIPCIDR, "clusterIP", svcIP)
+			klog.V(4).InfoS("Route with current ClusterIP CIDR can route the ClusterIP to Antrea gateway", "ClusterIP CIDR", curClusterIPCIDR, "ClusterIP", svcIP)
 		}
 	} else {
+		// Create the first ClusterIP route on Antrea gateway.
 		route := generateRoute(svcIP, mask, gw, linkIndex, scope)
 		if err := netlink.RouteReplace(route); err != nil {
 			return fmt.Errorf("failed to install new ClusterIP route: %w", err)
 		}
-
+		klog.V(4).InfoS("Created a route to route the ClusterIP to Antrea gateway", "ClusterIP", svcIP)
 		if isIPv6 {
 			c.clusterIPv6CIDR = route.Dst
 		} else {
 			c.clusterIPv4CIDR = route.Dst
 		}
+
+		// List all existing routes on Antrea gateway.
+		routes, err := c.listIPRoutesOnGW()
+		if err != nil {
+			return fmt.Errorf("error listing ip routes: %v", err)
+		}
+
+		for i := 0; i < len(routes); i++ {
```

Comment:
maybe the route cleanup logic could be shared between the 2 cases to simplify the code?

```go
if curClusterIPCIDR != nil && curClusterIPCIDR.Contains(svcIP) {
    return nil
}
newClusterIPCIDR := ... // depends on curClusterIPCIDR
route := generateRoute(newClusterIPCIDR.IP, mask, gw, linkIndex, scope)
netlink.RouteReplace(route)
// delete all unneeded routes
```

Reply:
How about this:

```go
func (c *Client) AddClusterIPRoute(svcIP net.IP) error {
	isIPv6 := utilnet.IsIPv6(svcIP)
	linkIndex := c.nodeConfig.GatewayConfig.LinkIndex
	scope := netlink.SCOPE_UNIVERSE
	curClusterIPCIDR := c.clusterIPv4CIDR
	mask := ipv4AddrLength
	gw := config.VirtualServiceIPv4
	if isIPv6 {
		curClusterIPCIDR = c.clusterIPv6CIDR
		mask = ipv6AddrLength
		gw = config.VirtualServiceIPv6
	}

	// If the route exists and contains the ClusterIP, it's no need to update the route.
	if curClusterIPCIDR != nil && curClusterIPCIDR.Contains(svcIP) {
		klog.V(4).InfoS("Route with current ClusterIP CIDR can route the ClusterIP to Antrea gateway", "ClusterIP CIDR", curClusterIPCIDR, "ClusterIP", svcIP)
		return nil
	}

	var newClusterIPCIDR *net.IPNet
	var err error
	if curClusterIPCIDR != nil {
		// If the route exists and doesn't contain the ClusterIP, generate a new destination ipNet.
		newClusterIPCIDR, err = util.ExtendCIDRWithIP(curClusterIPCIDR, svcIP)
		if err != nil {
			return fmt.Errorf("extend route CIDR with error: %w", err)
		}
	} else {
		// If the route doesn't exist, generate a new destination with the ClusterIP. Note that, this is the first ClusterIP
		// to install a route.
		newClusterIPCIDR = &net.IPNet{IP: svcIP, Mask: net.CIDRMask(mask, mask)}
	}

	// Generate a route with the new ClusterIP CIDR and install it.
	newClusterIPCIDRMask, _ := newClusterIPCIDR.Mask.Size()
	route := generateRoute(newClusterIPCIDR.IP, newClusterIPCIDRMask, gw, linkIndex, scope)
	if err = netlink.RouteReplace(route); err != nil {
		return fmt.Errorf("failed to install new ClusterIP route: %w", err)
	}
	if isIPv6 {
		c.clusterIPv6CIDR = route.Dst
	} else {
		c.clusterIPv4CIDR = route.Dst
	}
	klog.V(4).InfoS("Created a route to route the ClusterIP to Antrea gateway", "route", route, "ClusterIP", svcIP)

	// Collect stale routes.
	var staleRoutes []*netlink.Route
	if curClusterIPCIDR != nil {
		// If current ClusterIP CIDR is not nil, the route whose destination is current CIDR should be uninstalled.
		route.Dst = curClusterIPCIDR
		staleRoutes = []*netlink.Route{route}
	} else {
		// If current ClusterIP CIDR is nil, which means that Antrea Agent has just starts up, and all existing routes
		// whose destination contains the first ClusterIP should be uninstalled, except the newly installed ClusterIP route.
		// Note that, there may be multiple stale ClusterIP routes prior to this commit. When upgrading, all stale ClusterIP
		// routes will be collected. Of course, there will be only one stale ClusterIP route when Antrea Agent starts up
		// after this commit.
		routes, err := c.listIPRoutesOnGW()
		if err != nil {
			return fmt.Errorf("error listing ip routes: %w", err)
		}
		for i := 0; i < len(routes); i++ {
			if routes[i].Gw.Equal(gw) && !routes[i].Dst.IP.Equal(svcIP) && routes[i].Dst.Contains(svcIP) {
				staleRoutes = append(staleRoutes, &routes[i])
			}
		}
	}

	// Remove stale routes.
	for _, rt := range staleRoutes {
		if err = netlink.RouteDel(rt); err != nil {
			return fmt.Errorf("failed to uninstall stale ClusterIP route %s: %w", rt.String(), err)
		}
		klog.V(4).InfoS("Uninstalled stale ClusterIP route successfully", "stale route", rt)
	}

	return nil
}
```

Reply:
LGTM